### PR TITLE
added components to the yaml config.

### PR DESCRIPTION
**Pull request by dITySoftware**
… config.

### CODE REVIEW

**Review by whotwagner — CHANGES_REQUESTED** [review]
In addition I sent you an Email with another big change-request

File: .travis.yml [hunk 1]
```diff
@@ -21,6 +21,7 @@ install:
   - sudo pip3 install -U scipy
   - sudo ln -s /usr/lib/python3.6/site-packages/numpy /etc/aminer/conf-enabled/
   - sudo ln -s /usr/lib/python3.6/site-packages/scipy /etc/aminer/conf-enabled/
+  - sudo ln -s /usr/lib/python3.6/site-packages/cerberus /etc/aminer/conf-enabled/
```
[the remark this answers is not in the document]
Comment (dITySoftware) [reply .travis.yml]: removed links for cerberus scipy and numpy as they are in the ansible role.

File: aecid-testsuite/runAMinerDemo.sh [hunk 2]
```diff
@@ -1,4 +1,10 @@
-cp $1 /tmp/demo-config.py
+if [[ $1 == *.py ]]; then
+    cp $1 /tmp/demo-config.py
+elif [[ $1 == *.yml ]]; then
+    cp $1 /tmp/demo-config.yml
+else
+    exit 2
```
[the remark this answers is not in the document]
Comment (dITySoftware) [reply aecid-testsuite/runAMinerDemo.sh:8]: aminerDemo.sh only has two exit codes: 0 - success, 1 - failure.
Therfor it is clear what exit code 2 means. Is it necessary to echo an error message here?

These exit codes just signal Travis to fail. The scripts are not thought to be used by users.

File: aecid-testsuite/unit/data/configfiles/template_config.yml [hunk 3]
```diff
@@ -113,33 +19,32 @@ Parser:
...
 
 Analysis:
-        - type: "NewMatchPathValueDetector"
+        - type: NewMatchPathValueDetector
           paths: ["/model/accesslog/status"]
           persistence_id: 'accesslog_status' # optional default: Default
-          output_logline: false # optional default: true
-          learnMode: true
+          output_logline: False # optional default: True
+          auto_include_flag: True
```
[the remark this answers is not in the document]
Comment (dITySoftware) [reply aecid-testsuite/unit/data/configfiles/template_config.yml:31]: i refactored it to learn_mode as we are using snake mode for everything now. Is that OK?

File: aecid-testsuite/unit/config/YamlConfigTest.py [hunk 4]
```diff
@@ -18,45 +44,44 @@ def setUp(self):
...
-    """ This test builds a analysis_pipeline from a valid yaml-file """
-    def test_analysis_pipeline_working_config(self):
+    def test5_analysis_pipeline_working_config(self):
+        """This test builds a analysis_pipeline from a valid yaml-file."""
         spec = importlib.util.spec_from_file_location('aminer_config', '/usr/lib/logdata-anomaly-miner/aminer/ymlconfig.py')
         aminer_config = importlib.util.module_from_spec(spec)
         spec.loader.exec_module(aminer_config)
-        aminer_config.loadYaml('unit/data/configfiles/apache_inline_config.yml')
-        from aminer.AnalysisChild import AnalysisContext
+        aminer_config.load_yaml('unit/data/configfiles/apache_inline_config.yml')
```
[the remark this answers is not in the document]
Comment (dITySoftware) [reply aecid-testsuite/unit/config/YamlConfigTest.py:84]: renamed all apache configs to multiple_components.

File: aecid-testsuite/unit/config/YamlConfigTest.py [hunk 5]
```diff
@@ -103,67 +131,296 @@ def test_analysis_pipeline_working_config(self):
...
+        with self.assertRaises(ValueError):
+            aminer_config.load_yaml('unit/data/configfiles/unknown_event_handler.yml')
+
+    def test12_analysis_pipeline_working_config_without_event_handler_components(self):
+        """This test checks if the config can be loaded without any event handler components. This also tests if the
+        StreamPrinterEventHandler was loaded by default."""
+        spec = importlib.util.spec_from_file_location('aminer_config', '/usr/lib/logdata-anomaly-miner/aminer/ymlconfig.py')
+        aminer_config = importlib.util.module_from_spec(spec)
+        spec.loader.exec_module(aminer_config)
+        aminer_config.load_yaml('unit/data/configfiles/apache_inline_config_null_event_handlers.yml')
```
[the remark this answers is not in the document]
Comment (dITySoftware) [reply aecid-testsuite/unit/config/YamlConfigTest.py]: yes it is nessecary to load the config. However it is possible to delete one config and just remove 'EventHandlers' or 'Analysis' from the dictionary and reload the context. I have done that in commit 39bdf0e716709c7947fbcb94beb3f674a6d6793a.

File: source/root/usr/lib/logdata-anomaly-miner/aminer/ymlconfig.py [hunk 6]
```diff
@@ -253,53 +364,378 @@ def build_analysis_pipeline(analysis_context):
...
-            if item['type'] == 'StreamPrinterEventHandler':
+        if 'EventHandlers' in yaml_data and yaml_data['EventHandlers'] is not None:
+            for item in yaml_data['EventHandlers']:
+                func = item['type'].func
+                ctx = None
+                if item['type'].name in ('StreamPrinterEventHandler', 'DefaultMailNotificationEventHandler'):
+                    ctx = func(analysis_context)
+                if item['type'].name == 'SyslogWriterEventHandler':
+                    ctx = func(analysis_context, item['instance_name'])
+                if item['type'].name == 'KafkaEventHandler':
```
[the remark this answers is not in the document]
Comment (dITySoftware) [reply source/root/usr/lib/logdata-anomaly-miner/aminer/ymlconfig.py:716]: I have deleted it as most of the code is deprecated (for example args is no longer used).

I readded the code and change variables like learnMode to snake case.
Kafka is not tested yet. This needs to be done in future pull requests.

File: source/root/usr/lib/logdata-anomaly-miner/aminer/ymlconfig.py [hunk 7]
```diff
@@ -253,53 +364,378 @@ def build_analysis_pipeline(analysis_context):
...
+                    ctx = func(analysis_context, item['topic'], item['options'])
                 if item['json'] is True:
                     from aminer.events import JsonConverterHandler
-                    ctx = JsonConverterHandler([func(analysis_context)], analysis_context)
-                else:
+                    ctx = JsonConverterHandler([ctx], analysis_context)
```
[the remark this answers is not in the document]
Comment (dITySoftware) [reply source/root/usr/lib/logdata-anomaly-miner/aminer/ymlconfig.py:746]: no we can not. I have changed the order of the code, so the None check is performed before.

File: aecid-testsuite/unit/config/YamlConfigTest.py [hunk 8]
```diff
@@ -103,67 +131,296 @@ def test_analysis_pipeline_working_config(self):
...
         spec.loader.exec_module(aminer_config)
-        aminer_config.loadYaml('unit/data/configfiles/apache_inline_config_null_analysis_components.yml')
-        from aminer.AnalysisChild import AnalysisContext
+        aminer_config.load_yaml('unit/data/configfiles/apache_inline_config_null_analysis_components.yml')
         context = AnalysisContext(aminer_config)
         context.build_analysis_pipeline()
-        self.run_empty_analysis_components_tests(context)
+        self.run_empty_components_tests(context)
 
         spec = importlib.util.spec_from_file_location('aminer_config', '/usr/lib/logdata-anomaly-miner/aminer/ymlconfig.py')
```
[the remark this answers is not in the document]
Comment (dITySoftware) [reply aecid-testsuite/unit/config/YamlConfigTest.py]: I have actually changed the code like with 'EventHandlers' to not load a second config, but just deleting the entry in the aminer_config.yaml_data and reloading the context.

File: source/root/usr/lib/logdata-anomaly-miner/aminer/ymlconfig.py [hunk 9]
```diff
@@ -128,105 +192,152 @@ def build_analysis_pipeline(analysis_context):
...
+                    item['paths'],
+                    anomaly_event_handlers,
+                    auto_include_flag=learn,
                     persistence_id=item['persistence_id'],
                     output_log_line=item['output_logline'])
-            elif item['type'] == 'NewMatchPathValueComboDetector':
-                tmpAnalyser = func(
-                    analysis_context.aminer_config, item['paths'],
-                    anomaly_event_handlers, auto_include_flag=learn,
+            elif item['type'].name == 'MatchPathFilter':
```
[the remark this answers is not in the document]
Comment (dITySoftware) [reply source/root/usr/lib/logdata-anomaly-miner/aminer/ymlconfig.py:258]: template_config.py also does not contain all components. I think that is not the purpose of the template_config.
test17 covers the loading of the demo-config.py and demo-config.yml, which should be the same, and compared the contexts.

demo-config.yml contains all of the components.

Please confirm once again that the template_config also should contain all components.

File: aecid-testsuite/unit/data/configfiles/template_config.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (landauermax) [reply aecid-testsuite/unit/data/configfiles/template_config.yml]: I would say yes, it is ok in snake_case.

File: source/root/usr/lib/logdata-anomaly-miner/aminer/ymlconfig.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (whotwagner) [reply source/root/usr/lib/logdata-anomaly-miner/aminer/ymlconfig.py:258]: Good idea. Please make a comment in template_config.yml with a reference to demo-config.yml

**Review by whotwagner — CHANGES_REQUESTED** [review]

File: source/root/usr/lib/logdata-anomaly-miner/aminer/ymlconfig.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (dITySoftware) [reply source/root/usr/lib/logdata-anomaly-miner/aminer/ymlconfig.py:258]: Done.

**Review by whotwagner — CHANGES_REQUESTED** [review]
Just waiting for fixed travis-Issues

**Review by whotwagner — APPROVED** [review]

**Pull request merged**